pkg/availability/multisigcollector/internal/parts/certverification/certverification.go

```diff
@@ -4,103 +4,179 @@ import (
+type RequestID = uint64
+
+type RequestState struct {
+	certVerifiedValid map[t.BatchIDString]bool
+}
+
 // IncludeVerificationOfCertificates registers event handlers for processing availabilitypb.VerifyCert events.
 func IncludeVerificationOfCertificates(
 	m dsl.Module,
 	mc *common.ModuleConfig,
 	params *common.ModuleParams,
 	nodeID t.NodeID,
 ) {
+	state := State{
+		NextReqID:    0,
+		RequestState: make(map[RequestID]*RequestState),
+	}
 	// When receive a request to verify a certificate, check that it is structurally correct and verify the signatures.
-	adsl.UponVerifyCert(m, func(cert *apb.Cert, origin *apb.VerifyCertOrigin) error {
-		mscCert, err := verifyCertificateStructure(params, cert)
-		if err != nil {
-			adsl.CertVerified(m, t.ModuleID(origin.Module), err, origin)
+	apbdsl.UponVerifyCert(m, func(cert *apbtypes.Cert, origin *apbtypes.VerifyCertOrigin) error {
+
+		reqID := state.NextReqID
+		state.NextReqID++
+
+		mscCerts, err := verifyCertificateStructure(params, cert)
+		valid, errStr := t.ErrorPb(err)
+
+		if err != nil { // invalid certificate, send cert verified event with valid=False
+			apbdsl.CertVerified(m, origin.Module, valid, errStr, origin)
 			return nil
 		}
 
-		sigMsg := common.SigData(params.InstanceUID, t.BatchID(mscCert.BatchId))
-		dsl.VerifyNodeSigs(m, mc.Crypto,
-			/*data*/ sliceutil.Repeat(sigMsg, len(mscCert.Signers)),
-			/*signatures*/ mscCert.Signatures,
-			/*nodeIDs*/ t.NodeIDSlice(mscCert.Signers),
-			/*context*/ &verifySigsInCertContext{origin},
-		)
+		state.RequestState[reqID] = &RequestState{
+			certVerifiedValid: make(map[t.BatchIDString]bool),
+		}
+		nonEmptyCerts := make([]*mscpbtypes.Cert, 0)
+
+		for _, mscCert := range mscCerts {
+			if !emptycert.IsEmpty(mscCert) {
+				nonEmptyCerts = append(nonEmptyCerts, mscCert) // do not add to requests
+			}
+		}
+		if len(nonEmptyCerts) > 0 {
+			for _, mscCert := range nonEmptyCerts {
+				state.RequestState[reqID].certVerifiedValid[t.BatchIDString(mscCert.BatchId)] = false
+				sigMsg := common.SigData(params.InstanceUID, mscCert.BatchId)
+				dsl.VerifyNodeSigs(m, mc.Crypto,
+					/*data*/ sliceutil.Repeat(sigMsg, len(mscCert.Signers)),
+					/*signatures*/ mscCert.Signatures,
+					/*nodeIDs*/ mscCert.Signers,
+					/*context*/ &verifySigsInCertContext{origin, reqID, mscCert},
+				)
+			}
+		} else {
+			// all certs are empty cert, verify immediately
+			apbdsl.CertVerified(m, origin.Module, true, "", origin)
+			delete(state.RequestState, reqID)
+		}
+
 		return nil
 	})
 
 	// When the signatures in a certificate are verified, output the result of certificate verification.
 	dsl.UponNodeSigsVerified(m, func(nodeIDs []t.NodeID, errs []error, allOK bool, context *verifySigsInCertContext) error {
+		reqID := context.reqID
+
+		if _, ok := state.RequestState[reqID]; !ok {
+			return nil
+		}
+
+		state.RequestState[reqID].certVerifiedValid[t.BatchIDString(string(context.cert.BatchId))] = allOK
```

Comment:
The string cast seems to be superfluous here (at least my IDE says so).

Reply:
done